checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -111,6 +119,84 @@ public boolean isValidUse(
         return ret;
     }
 
+    // For assignments and local variable initialization:
+    // Check for @UI annotations on the lhs, use them to infer @UI types on lambda expressions in the rhs
+    private void lambdaAssignmentCheck(
+            AnnotatedTypeMirror varType, LambdaExpressionTree lambdaExp, String errorKey) {
+        AnnotatedTypeMirror lambdaType = atypeFactory.getAnnotatedType(lambdaExp);
+        assert lambdaType != null : "null type for expression: " + lambdaExp;
+        commonAssignmentCheck(varType, lambdaType, lambdaExp, errorKey);
+    }
+
+    @Override
+    public Void visitVariable(VariableTree node, Void p) {
+        Void result = super.visitVariable(node, p);
+        if (node.getInitializer() != null
+                && node.getInitializer().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
+            lambdaAssignmentCheck(
+                    atypeFactory.getAnnotatedType(node),
+                    (LambdaExpressionTree) node.getInitializer(),
+                    "assignment.type.incompatible");
+        }
+        return result;
+    }
+
+    @Override
+    public Void visitAssignment(AssignmentTree node, Void p) {
+        Void result = super.visitAssignment(node, p);
+        if (node.getExpression().getKind() == Tree.Kind.LAMBDA_EXPRESSION) {
+            lambdaAssignmentCheck(
+                    atypeFactory.getAnnotatedType(node.getVariable()),
+                    (LambdaExpressionTree) node.getExpression(),
+                    "assignment.type.incompatible");
+        }
+        return result;
+    }
+
+    // Returning a lambda expression also triggers inference based on the method's return type.
+    @Override
+    public Void visitReturn(ReturnTree node, Void p) {
+        Void result = super.visitReturn(node, p);
+        /*if (node.getExpression() != null && TreeUtils.enclosingClass(getCurrentPath()).toString().contains("Java8"))
+        System.err.println(TreeUtils.enclosingMethodOrLambda(getCurrentPath()) + " returns: " + node.getExpression() + " with type " +
+            atypeFactory.getAnnotatedType(node.getExpression()));*/
```

Comment:
This code above, when uncommented, shows that for `Java8Lambdas.returnLambdasTest2`, the checker indeed knows the type of the formal is `@AlwaysSafe` and the actual parameter is `@AlwaysSafe @UI`, but `lambdaAssignmentCheck` (and, by extension, `commonAssignmentCheck`) fails to report an error.

Reply:
```@AlwaysSafe @UI T``` isn't a valid type for any ```T```.  Something ought to be complaining, unless you're explicitly adding one of those qualifiers.